specs/capri/OpenAPI_Design.md

```diff
@@ -0,0 +1,160 @@
+To achieve various purposes, VMware supports multi types of plugins. All of them rely on storage APIs to interact with backend storage system.
+So if we implement a common framework that encapsulates API details of backend storage and provision more generic interfaces, it would be much more easier for plugin developers to implement a plugin, because developers can concern more to plugin itself's logic, not stuck in communication with storage.
+
+### Goals
+
+* Provide northbound generic APIs that encapsulate storage northbound APIs(through REST or others).
+* Open scalable&pluggable framework, 3rd parties can encapsulate its own storage's APIs following northbound interfaces' definition.
+
+### Non-Goals
+
+None.
+
+## Design Details
+
+### OpenAPI Usage Architecture
+
+![](media/OpenAPI_Usage_Architecture.png)
+OpenAPI is a language dependant library, such as a Jar package for Java.
+VMware plugins should import the OpenAPI library and call its northbound interfaces to communicate with corresponding storage.
+
+### OpenAPI Module
+
+![](media/OpenAPI_Module.png)
+OpenAPI provides unified northbound interfaces to VMware plugins, no matter what southbound storages connected.
+This could filter the differences between various of storages. Here we also see OpenSDS as a storage.
+To achieve the storage-not-aware purpose, each storage vendor must provide its own plugin to OpenAPI, which encapsulates storage calls based on storage's API definition.
+OpenAPI's northbound interfaces invokes these plugins' functions to actually interact with storages.
+
+### Data model impact
+
+No.
+
+### REST API impact
+
+May require OpenSDS to provide more REST APIs.
+
+### Security impact
+
+No.
+
+### Other end user impact
+
+No.
+
+### Performance impact
+
+No.
+
+### Other deployer impact
+
+No.
+
+### Developer impact
+
+Developers can participate in OpenAPI to implement more functions, so that OpenAPI can provide more interfaces to VMware plugins.
+Or developers can commit 3rd-party plugin to scale OpenAPI up to other storages.
+
+## Use Cases
+
+For OpenAPI, various of VMware plugins are users.
+Whenever VMware plugins want to operate with storages, they invoke unified interfaces of OpenAPI.
+
+## Implementation
+
+### DeviceFactory Definition
+
+*DeviceFactory* is a factory that responsible for plugin loading and initializing, according to what storage VMware plugin want to use.
+
+### Device Northbound Interface Definitions
+
+For this very first version of OpenAPI, we only supply the most basic interfaces that VMware plugins might use.
+Each plugin is supposed to inherit these interfaces and implement accordingly.
+Further functions will be gradually added based on requirements later.
+
+| Interface | Description |
+|--|--|
+| login | Login to storage |
```

Comment:
Should these interfaces be exposed to the public? If so, please use capital letter instead.